ibis/sql/postgres/tests/test_udf.py

```diff
@@ -0,0 +1,173 @@
+        schema_exists = df_result['n_recs'].iloc[0] > 0
+        tries += 1
+    assert tries > 0
+    global test_schema
+    test_schema = new_name
+    connection.con.execute("CREATE SCHEMA {};".format(new_name))
+
+
+table_name = 'udf_test_users'
+
+sql_table_setup = """DROP TABLE IF EXISTS {schema}.{table_name};
+CREATE TABLE {schema}.{table_name} (
+    user_id integer,
+    user_name varchar,
+    name_length integer
+);
+INSERT INTO {schema}.{table_name} VALUES
+(1, 'Raj', 3),
+(2, 'Judy', 4),
+(3, 'Jonathan', 8)
+;
+"""
+
+sql_define_py_udf = """CREATE OR REPLACE FUNCTION {schema}.pylen(x varchar)
+RETURNS integer
+LANGUAGE plpythonu
+AS
+$$
+return len(x)
+$$;"""
+
+sql_define_udf = """CREATE OR REPLACE FUNCTION {schema}.custom_len(x varchar)
+RETURNS integer
+LANGUAGE SQL
+AS
+$$
+SELECT length(x);
+$$;"""
+
+
+@pytest.fixture
+def con_for_udf(con):
+    create_test_schema(con)
+    con.con.execute(
+        sql_table_setup.format(table_name=table_name, schema=test_schema)
+    )
+    con.con.execute(sql_define_udf.format(schema=test_schema))
+    con.con.execute(sql_define_py_udf.format(schema=test_schema))
+    yield con
+    # teardown
+    con.con.execute("DROP SCHEMA IF EXISTS {} CASCADE".format(test_schema))
+
+
+@pytest.fixture
+def table(con_for_udf):
+    return con_for_udf.table(table_name, schema=test_schema)
+
+# Tests
+
+
+def test_sql_length_udf_worked(table):
+    """Test creating ibis UDF object based on existing UDF in the database"""
+    # Create ibis UDF objects referring to UDFs already created in the database
+    custom_length_udf = ibis.sql.postgres.udf.api.existing_udf(
+        'custom_len',
+        input_types=[ibis.expr.datatypes.String()],
+        output_type=ibis.expr.datatypes.Integer(),
+        schema=test_schema
+    )
+    result_obj = table[
+        table,
+        custom_length_udf(table['user_name']).name('custom_len')
+    ]
+    result = result_obj.execute()
+    assert result['custom_len'].sum() == result['name_length'].sum()
+
+
+def test_py_length_udf_worked(table):
+    # Create ibis UDF objects referring to UDFs already created in the database
+    py_length_udf = ibis.sql.postgres.udf.api.existing_udf(
+        'pylen',
+        input_types=[ibis.expr.datatypes.String()],
+        output_type=ibis.expr.datatypes.Integer(),
+        schema=test_schema
+    )
+    result_obj = table[
+        table,
+        py_length_udf(table['user_name']).name('custom_len')
+    ]
+    result = result_obj.execute()
+    assert result['custom_len'].sum() == result['name_length'].sum()
+
+
+def mult_a_b(a, b):
+    """Test function to be defined in-database as a UDF
+    and used via ibis UDF"""
+    return a * b
+
+
+def test_func_to_udf_smoke(con_for_udf, table):
+    """Test creating a UDF in database based on Python function
+    and then creating an ibis UDF object based on that"""
+    mult_a_b_udf = ibis.sql.postgres.udf.api.func_to_udf(
```

Comment:
I'm not a huge fan of this API. What do you think about this?

```python
@ibis_client.udf([dt.int32, dt.int32], dt.int32, schema='my_schema', overwrite=True)
def mult_a_b(a, b):
    return a * b
```

Reply:
I'm totally open to changing the API. I was so focused on just getting something that worked, and was actually hoping you and the community would chime in on how to improve the API. I'll work on refactoring. 

What would you prefer the API be for existing UDFs? Still have a function like the `existing_udf` that I have? Would you prefer a different name for it?

Reply:
My most recent push [47aacdd7](https://github.com/scottcode/ibis/commit/47aacdd756a70eb9d81374b73ff4e212ce0d8d01) adds your preferred `@ibis_client.udf(...)` decorator API.